Paper: PyQtGraph - High Performance Visualization for All Platforms

ntjess opened the pull request:
👋 Excited to attend SciPy in person this year!

Tagging first author @j9ac9k to get pinged as well for updates

---

If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@mepa<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@saulshanabrook<!--end-reviewers-list-->

Note: Waiting on #827 for code block highlighting

ntjess committed:
add pyqtgraph paper

deniederhut commented:
@scoobies assign @cbcunc as editor

scoobies commented:
Assigned! @cbcunc is now the editor

deniederhut commented:
@scoobies add @yashasvimisra2798 to reviewers

scoobies commented:
@yashasvimisra2798 added to the reviewers list!

deniederhut commented:
@scoobies code of conduct

scoobies commented:
Our CoC: https://www.scipy2023.scipy.org/code-of-conduct

scoobies commented:
Reports of abusive or harassing behavior may be reported to scipy@enthought.com

cbcunc commented:
Editor says paper is complete and builds. Awaiting review.

cbcunc commented:
@scoobies remove editor

scoobies commented:
Editor removed!

cbcunc commented:
@scoobies assign @mepa as editor

scoobies commented:
Assigned! @mepa is now the editor

mepa commented:
@scoobies add @saulshanabrook to reviewers

scoobies commented:
@saulshanabrook added to the reviewers list!

saulshanabrook commented:
I am looking forward to reviewing this paper! I will likely get to it early next week.

mepa commented:
Hey @yashasvimisra2798 and @saulshanabrook! A friendly reminder that initial reviews are due Friday, June 23rd.

saulshanabrook commented on papers/moore-pyqtgraph/metadata.yaml:
```diff
@@ -0,0 +1,37 @@
+---
+abstract:
+  - PyQtGraph is a plotting library with high performance, cross-platform support and interactivity as its primary objectives. These goals are achieved by connecting the Qt GUI framework and the scientific Python ecosystem. The end result is a plotting library that supports the use of python and NumPy arrays to drive interactive visualizations on all major operating systems.
```
I was unclear if you meant you have supported for NumPy arrays as well as the builtin `array` module in Python. I assumed you just mean NumPy, but if not it might be helpful to clarify.

```suggestion
  - PyQtGraph is a plotting library with high performance, cross-platform support and interactivity as its primary objectives. These goals are achieved by connecting the Qt GUI framework and the scientific Python ecosystem. The end result is a plotting library that supports the use of NumPy arrays to drive interactive visualizations on all major operating systems.
```

saulshanabrook replied:
Thanks for making this clearer to me!

saulshanabrook commented on papers/moore-pyqtgraph/paper.tex:
```diff
@@ -0,0 +1,285 @@
+The supplementary information includes video demonstrations of two additional applications that make heavy use of PyQtGraph functionality to explore spectral data and to visualize volumetric data representing the 3d structure of multilayer circuit boards.
+
+\section{Software development}
+PyQtGraph was first released in 2012, under the open source MIT license. It is known to run on systems ranging from the Raspberry Pi to IBM's s390x architecture. Development is coordinated by volunteer maintainers, with additional code provided by occasional contributors. A continuous integration system asserts that the codebase passes a suite of tests for different combinations of Qt bindings, P ...
```
I would be curious to hear more about the original impetus for creating the library, to gain more context into the design decisions.

ntjess replied:
@campagnola are you able to add a few sentences about your initial motivation for this library? You can also just reply to this comment and I'm happy to update the paper if that's easier👍

campagnola replied:
The original motivation for pyqtgraph was in data acquisition software, where there is a need to be able to display video and plots with realtime frame rates and interactivity that allows data exploration. Interactivity is an important point here, because we already had matplotlib which is excellent for visualizing data in a way that tells a particular story. When you are looking at new data, though, it doesn't have a story yet. You want to be able to slice it and stretch it and look at it from every possible angle, quickly and easily, so that you can decide what story to tell about it. 

At the time, most acquisition software would have been written in C/C++ for efficiency, and we had just begun to cross the threshold where python + numpy + Qt together were fast enough. We also had PyQwt which was perfect for this purpose, but went through a long period without a maintainer (I presume because, at the time, it was a huge burden maintaining and distributing compiled python packages). So pyqtgraph began as a replacement for PyQwt that would be pure-python, and thus easier to develop and distribute. Following that template, it was also to include UI elements that have common use in acquisition/analysis applications, but are missing from Qt (for example, tools for adjusting image contrast, parameter trees, etc.).

saulshanabrook replied:
Thank you for the details, that helps a lot to understand the motivations.

saulshanabrook commented on papers/moore-pyqtgraph/paper.tex:
```diff
@@ -0,0 +1,285 @@
+
+\subsubsection{Parameter Trees}\label{sec:paramtrees}
+% \color{DarkOrchid}
+Another common requirement for user interaction is a a mechanism to interact with for configuration settings or algorithm parameters. PyQtGraph provides this capability through the \texttt{ParameterTree} object. This takes a simple Python dictionary describing the parameters and leverages existing Qt widgets for editing of text, numeric, list-like, and custom data types. Parameters can be grouped ...
```
It's not needed, but if you had any comparison to other Python libraries which provide similar models of configuration management, such as traitlets, I would be curious to hear them.

saulshanabrook replied:
Thank you for adding more details here! It is clear now to me how it compares:

```
Another common requirement for user interaction is a mechanism to interact with for configuration settings or algorithm parameters. PyQtGraph provides this capability through the \texttt{ParameterTree} object which hosts any number of \texttt{Parameter}s. Similar to \texttt{traitlets}, PyQtGraph's \texttt{Parameter} objects encapsulate a value type and allow registering callbacks, performing input validation, and more. However, \texttt{Parameter}s are different in that most are coupled to a widget representation, i.e., allowing users to easily update the values graphically.
\texttt{Parameter} objects can be created through a simple Python dictionary listing its specifications (type, value, and traits such as `readonly' or value range). It is then bound to a Qt widget for editing text, numeric, list-like, and custom data depending on the parameter type. Parameters can be grouped, linked, and dynamically instantiated or removed. Callbacks for user actions or value changes allow results to be recalculated immediately (i.e., while a spinbox is changing its value) or after the value has settled. Parameter trees can save and load their states hierarchically to easily create persistent configuration files. Since accessing parameters mimics a Python dictionary, they can function as a drop-in replacement for programmatically adjusted settings rather than forcing users to interface through widgets alone.
```

saulshanabrook commented on papers/moore-pyqtgraph/paper.tex:
```diff
@@ -0,0 +1,285 @@
+
+
+\subsection{Implementation}
+\texttt{GraphicsView} renders line segments in a freely scaled coordinate system through \texttt{QPainterPath} objects. The rendering performance of PyQtGraph results from optimized code to create such paths directly from NumPy ndarrays describing sets of $x$ and $y$ coordinates. The core function \texttt{arrayToQPath} uses NumPy's structured array functionality to efficiently create a binary com ...
```
After I read this paper one question I had was how it might compare to in-browser interactive visualization systems in Python. There are many different ones to choose from (Altair, Plotly, Bokeh...).

My assumption is that by using a native tool like QT you can move the data directly without having to copy it or send it to the browser, which is essential for the type of performance you are getting. If so, it might be helpful to articulate this explicitly.

saulshanabrook reviewed (approved):
I apologize for the late review!

Overall, I found the paper interesting and I had a good sense of the package by reading it, without any prior knowledge beforehand.

My main feedback is that I would be interested in hearing some more comparisons between different tools in the same as well as a more explicit story about where the package came from and what needs it was created to solve. This would help me understand the larger narrative around it. But it is not essential to the paper as a whole, which I think does a good job of capturing the scope and some applications.

mepa commented:
@scoobies remove @yashasvimisra2798 from reviewers

scoobies commented:
@yashasvimisra2798 removed from the reviewers list!

mepa commented:
@scoobies mark pending comment

ntjess committed:
Address reviewer comment:
Link: https://github.com/scipy-conference/scipy_proceedings/pull/828#discussion_r1243958767

ntjess committed:
Incorporate PyQtGraph's backstory
Link: https://github.com/scipy-conference/scipy_proceedings/pull/828#discussion_r1256512614

ntjess committed:
Add script to cycle through examples

ntjess committed:
Address reviewer comment about parameter trees.
A more detailed comparison against `traitlets` and other python libraries for parameter management is justifiable, but
unfortunately not tackled here due to time constraints
Link to reviewer comment: https://github.com/scipy-conference/scipy_proceedings/pull/828#discussion_r1243962611

ntjess committed:
add poster material

ntjess committed:
remove unused crosshair.png

ntjess committed:
add info.json for zenodo

deniederhut commented:
Hi @ntjess ! Is this paper ready for re-review?

ntjess commented:
Hi @deniederhut, I have one more set of changes (updated benchmark plots + acknowledgement), I plan to push shortly after the lightning talks if that's not too late

ntjess committed:
Several updates:
- Simplify code block illustrations
- Update benchmarks to more recent library version (Thanks @j9ac9k!)
- Grammar and wording review
- Update author order per author requests
- Add acknowledgements

ntjess commented:
>  I plan to push shortly after the lightning talks if that's not too late

@deniederhut Perhaps stretching the definition of shortly 😅 but authors are happy with the revision as of now. Thanks!

j9ac9k commented:
Hi @mepa 

I just checked the build server https://procbuild.scipy.org/# and when clicking on view PDF, it was the original submission w/o updates made in this PR.  Is that expected or is there an issue?

Thanks!

saulshanabrook commented:
@j9ac9k I hit the little rebuild icon on that site and it seemed to rebuild a recent copy!

I just looked through the changes and this looks great! Thank you for adding some more context. Congratulations on the wonderful paper and library.

j9ac9k commented:
> @j9ac9k I hit the little rebuild icon on that site and it seemed to rebuild a recent copy!
> 
> I just looked through the changes and this looks great! Thank you for adding some more context. Congratulations on the wonderful paper and library.

I had thought I tried that, but I can rebuild locally should I need to, as long as you were able to read an up to date copy, I'm good :D Thank you for the kind words and for your time reviewing our paper!

mepa commented:
> Hi @mepa
> 
> I just checked the build server https://procbuild.scipy.org/# and when clicking on view PDF, it was the original submission w/o updates made in this PR. Is that expected or is there an issue?
> 
> Thanks!

Hi @j9ac9k, there are two ways to trigger a paper rebuild on the build server:
- By commenting `@scoobies build paper` here in the PR.
- By clicking the "rebuild icon" next your paper at https://procbuild.scipy.org/#, as @saulshanabrook mentioned above.

Does the [current pdf](https://procbuild.scipy.org/download/ntjess-moore-pyqtgraph) look as expected?

j9ac9k commented:
> > Hi @mepa
> 
> > 
> 
> > I just checked the build server https://procbuild.scipy.org/# and when clicking on view PDF, it was the original submission w/o updates made in this PR. Is that expected or is there an issue?
> 
> > 
> 
> > Thanks!
> 
> 
> 
> Hi @j9ac9k, there are two ways to trigger a paper rebuild on the build server:
> 
> - By commenting `@scoobies build paper` here in the PR.
> 
> - By clicking the "rebuild icon" next your paper at https://procbuild.scipy.org/#, as @saulshanabrook mentioned above.
> 
> 
> 
> Does the [current pdf](https://procbuild.scipy.org/download/ntjess-moore-pyqtgraph) look as expected?

It does! Must have been user error,I apologize for the noise.

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3758/BF03201852 is OK
- 10.1145/3386331 is OK
- 10.1049/ir:20030905 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1109/TIM.2019.2914711 is OK
- 10.1177/2211068214553022 is OK
- 10.3389/fninf.2014.00003 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1214/aos/1043351250 is INVALID because of 'https://doi.org/' prefix
- DOI:10.1007/978-3-540-33037-0\_2 URL is INVALID
- 10.1088/1742-6596/898/10/112002 is INVALID
```

mepa commented:
Hi @ntjess and @j9ac9k - could you please check and resolve the `INVALID DOIs` marked by scoobies above? It looks like the first one or two need to have prefixes removed, and the third one should be `10.1088/1742-6596/898/11/112002`.

ntjess committed:
Address DOI errors

Ref: https://github.com/scipy-conference/scipy_proceedings/pull/828#issuecomment-1646563223

ntjess commented:
@mepa sorry for the delay, these should be addressed in this most recent push.

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1214/aos/1043351250 is OK
- 10.3758/BF03201852 is OK
- 10.1145/3386331 is OK
- 10.1049/ir:20030905 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1109/TIM.2019.2914711 is OK
- 10.1177/2211068214553022 is OK
- 10.1088/1742-6596/898/11/112002 is OK
- 10.3389/fninf.2014.00003 is OK

MISSING DOIs

- None

INVALID DOIs

- DOI:10.1007/978-3-540-33037-0_2 is INVALID (failed connection)
```

mepa commented on papers/moore-pyqtgraph/mybib.bib:
```diff
@@ -0,0 +1,215 @@
+    inseries = {Springer Handbooks Comp.Statistics},
+    publisher = {Springer, Berlin, Heidelberg},
+    year = {2008},
+    doi = { DOI: 10.1007/978-3-540-33037-0_2 },
```
```suggestion
    doi = { 10.1007/978-3-540-33037-0_2 },
```

mepa commented:
Hey @ntjess and @j9ac9k, if you could please make the one last little change to remove "DOI: " as suggested above, then we can mark the paper ready for publication.

ntjess committed:
Resolve last doi issue

Co-authored-by: Meghann Agarwal <4421561+mepa@users.noreply.github.com>

ntjess commented:
My apologies for the delay, it should be updated now. Thanks for the reminder!

mepa commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ntjess-moore-pyqtgraph scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1214/aos/1043351250 is OK
- 10.3758/BF03201852 is OK
- 10.1145/3386331 is OK
- 10.1049/ir:20030905 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.1109/TIM.2019.2914711 is OK
- 10.1177/2211068214553022 is OK
- 10.1007/978-3-540-33037-0_2 is OK
- 10.1088/1742-6596/898/11/112002 is OK
- 10.3389/fninf.2014.00003 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

mepa commented:
Thanks, @ntjess !

mepa commented:
@scoobies mark ready

deniederhut reviewed (approved):
Thanks @saulshanabrook for helping us get this paper ready for publication. And thanks @ntjess for joining us at SciPy this year!

![source-2935940343](https://github.com/scipy-conference/scipy_proceedings/assets/6868396/e1fe6177-b9bd-4d25-a6ab-414fbfe3b096)
